multicore_runtime/wsgi_config.py

```diff
@@ -0,0 +1,216 @@
+
+
+def get_module_config(filename):
+  """Returns the parsed module config."""
+  with open(filename) as f:
+    return appinfo_includes.Parse(f)
+
+
+def app_for_script(script):
+  """Returns the WSGI app specified in the input string, or None on failure."""
+  try:
+    app, unused_filename, err = wsgi.LoadObject(script)
+    if err:
+      raise err
+  except ImportError as e:
+    # Despite nominally returning an error object, LoadObject will sometimes
+    # just result in an exception. By raising the err object if it exists, we
+    # catch either.
+    logging.exception('Failed to import %s', script)
+    return None
+  else:
+    return app_wrapped_in_user_middleware(app)
+
+
+def app_wrapped_in_user_middleware(app):
+  """Returns the input WSGI app, wrapped in appengine_config middleware."""
+  add_middleware = get_add_middleware_from_appengine_config()
+  if add_middleware:
+    return add_middleware(app)
+  else:
+    return app
+
+
+def get_add_middleware_from_appengine_config():
+  """Tries to import appengine_config and return middleware; fails silently.
+
+  `env_config` is optionally part of GAE-compatible user code.
+  If the user chooses to define the `webapp_add_wsgi_middleware` function
+  there, then we will use it to wrap app scripts. This is used for e.g.
+  appstats. Many user apps do not have or need env_config, so failure
+  to import it or failure to find `webapp_add_wsgi_middleware` is not an
+  error.
+
+  Returns:
+    The appengine_config.webapp_add_wsgi_middleware function or None.
+  """
+  try:
+    import appengine_config  # pylint: disable=g-import-not-at-top
+    return getattr(appengine_config, 'webapp_add_wsgi_middleware', None)
+  except ImportError:
+    return None
+
+
+def static_app_for_handler(handler):
+  """Returns a WSGI app that serves static files as directed by the handler.
+
+  Args:
+    handler: An individual handler from appinfo_external.handlers
+      (appinfo.URLMap)
+
+  Returns:
+    A static file-serving WSGI app closed over the handler information.
+  """
+  url_re = handler.url  # handler.url is a regex, despite the name.
+  files = handler.static_files
+  upload = handler.upload
+  if not files:
+    if handler.static_dir:
+      # If static_files is not set, convert static_dir to static_files and also
+      # modify the url regex accordingly. See the appinfo.URLMap docstring for
+      # more information.
+      url_re = static_dir_url(handler)
+      files = '{dir}{backref}'.format(dir=handler.static_dir, backref=r'/\1')
+      upload = '{dir}{re}'.format(dir=handler.static_dir, re='/.*')
+    else:
+      # Neither static_files nor static_dir is set; log an error and return.
+      logging.error('No script, static_files or static_dir found for %s',
+                    handler)
+      return None
+  return static_app_for_regex_and_files(url_re, files, upload,
+                                        mime_type=handler.mime_type)
+
+
+def static_dir_url(handler):
```

Comment:
And yes, I know `handler.url` might be out of our control, but at least if you refer to it as `.format(url_re=handler.url)` that already helps a bit, and naming functions and variables accordingly makes it a lot better too.